dropshot/src/server.rs

```diff
@@ -149,15 +201,208 @@ impl<C: ServerContext> HttpServerStarter<C> {
+}
+
+impl<C: ServerContext> InnerHttpsServerStarter<C> {
+    fn start(self) -> HttpServer<C> {
+        let app_state = self.app_state;
+        let server = self.server;
+        let local_addr = self.local_addr;
+
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        let log_close = app_state.log.new(o!());
+        let graceful = server.with_graceful_shutdown(async move {
+            rx.await.expect(
+                "dropshot server shutting down without invoking close()",
+            );
+            info!(log_close, "received request to begin graceful shutdown");
+        });
+
+        let join_handle = tokio::spawn(async { graceful.await });
+
+        HttpServer {
+            app_state,
+            local_addr,
+            join_handle: Some(join_handle),
+            close_channel: Some(tx),
+        }
+    }
+
+    pub fn new(
+        config: &ConfigDropshot,
+        api: ApiDescription<C>,
+        private: C,
+        log: &Logger,
+    ) -> Result<InnerHttpsServerStarter<C>, GenericError> {
+        let app_state = Arc::new(DropshotState {
+            private,
+            config: ServerConfig {
+                request_body_max_bytes: config.request_body_max_bytes,
+                page_max_nitems: NonZeroU32::new(10000).unwrap(),
+                page_default_nitems: NonZeroU32::new(100).unwrap(),
+            },
+            router: api.into_router(),
+            log: log.new(o!()),
+        });
+
+        for (path, method, _) in &app_state.router {
+            debug!(app_state.log, "registered endpoint";
+                "method" => &method,
+                "path" => &path
+            );
+        }
+
+        let (conn, local_addr) = {
+            let dont_use_this_cert = load_certs(
+                "/Users/gk/oxide/src/dropshot/dropshot/data/\
+                 dont_use_this_cert.pem",
+            )?;
+            let dont_use_this_key = load_private_key(
+                "/Users/gk/oxide/src/dropshot/dropshot/data/dont_use_this_key.\
+                 rsa",
+            )?;
+            let mut cfg =
+                rustls::ServerConfig::new(rustls::NoClientAuth::new());
+            cfg.set_single_cert(dont_use_this_cert, dont_use_this_key)?;
+            cfg.set_protocols(&[b"h2".to_vec(), b"http/1.1".to_vec()]);
+            let cfg = Arc::new(cfg);
+            let acceptor = TlsAcceptor::from(cfg);
+
+            let tcp = TcpListener::from_std(std::net::TcpListener::bind(
+                &config.bind_address,
+            )?)?;
+
+            let local_addr = tcp.local_addr()?;
+
+            (
+                stream! {
+                    loop {
+                        let (socket, addr) = tcp.accept().await?;
+                        let stream = acceptor.accept(socket).await?;
+                        let conn = TlsConn::new(stream, addr);
+                        yield Ok(conn)
+                    }
+                },
+                local_addr,
+            )
+        };
```

Comment:
I'm not sure I grok the need for all the `TlsConn`  + `HttpAcceptor` structures.

If we have a `TcpListener` object (`tcp`) that is configured to use HTTPS, couldn't we just use:

```rust
let stream = hyper::server::conn::AddrIncoming::from_listener(tcp)?;
let server = Server::builder(stream);
```

Reply:
Nvm - I see I got my configs mixed up